crates/bevy_pbr/src/render/mesh.rs

```diff
@@ -106,81 +107,52 @@ bitflags::bitflags! {
-        ),
-        Without<NotShadowCaster>,
-    >,
-    not_caster_query: Query<
-        (
-            Entity,
-            &ComputedVisibility,
-            &GlobalTransform,
-            &Handle<Mesh>,
-            Option<&NotShadowReceiver>,
-        ),
-        With<NotShadowCaster>,
-    >,
+    mut prev_len_caster: Local<usize>,
+    mut prev_len_not: Local<usize>,
+    caster_query: Query<ExtractMeshQuery, Without<NotShadowCaster>>,
+    not_caster_query: Query<ExtractMeshQuery, With<NotShadowCaster>>,
 ) {
-    let mut caster_values = Vec::with_capacity(*previous_caster_len);
-    for (entity, computed_visibility, transform, handle, not_receiver) in caster_query.iter() {
-        if !computed_visibility.is_visible {
-            continue;
-        }
+    let mesh_bundle = |(entity, _, transform, handle, not_receiver): ExtractMeshItem| {
         let transform = transform.compute_matrix();
-        caster_values.push((
-            entity,
-            (
-                handle.clone_weak(),
-                MeshUniform {
-                    flags: if not_receiver.is_some() {
-                        MeshFlags::empty().bits
-                    } else {
-                        MeshFlags::SHADOW_RECEIVER.bits
-                    },
-                    transform,
-                    inverse_transpose_model: transform.inverse().transpose(),
-                },
-            ),
-        ));
-    }
-    *previous_caster_len = caster_values.len();
-    commands.insert_or_spawn_batch(caster_values);
-
-    let mut not_caster_values = Vec::with_capacity(*previous_not_caster_len);
-    for (entity, computed_visibility, transform, mesh, not_receiver) in not_caster_query.iter() {
-        if !computed_visibility.is_visible {
-            continue;
-        }
-        let transform = transform.compute_matrix();
-        not_caster_values.push((
-            entity,
-            (
-                mesh.clone_weak(),
-                MeshUniform {
-                    flags: if not_receiver.is_some() {
-                        MeshFlags::empty().bits
-                    } else {
-                        MeshFlags::SHADOW_RECEIVER.bits
-                    },
-                    transform,
-                    inverse_transpose_model: transform.inverse().transpose(),
-                },
-                NotShadowCaster,
-            ),
-        ));
-    }
-    *previous_not_caster_len = not_caster_values.len();
-    commands.insert_or_spawn_batch(not_caster_values);
+        let flags = if not_receiver.is_some() {
+            MeshFlags::empty().bits
+        } else {
+            MeshFlags::SHADOW_RECEIVER.bits
+        };
+        let uniform = MeshUniform {
+            flags,
+            transform,
+            inverse_transpose_model: transform.inverse().transpose(),
+        };
+        (entity, (handle.clone_weak(), uniform))
+    };
+    let with_marker = |(entity, (handle, uniform))| (entity, (handle, uniform, NotShadowCaster));
```

Comment:
```suggestion
    let  add_marker_component = |(entity, (handle, uniform))| (entity, (handle, uniform, NotShadowCaster));
```

This is a closure, so more verb-like names are much clearer.